### PR TITLE
[run_lm_finetuning] Train from scratch

### PR DESCRIPTION
**Pull request by julien-c**
Ability to train a model from scratch, rather than finetune a pretrained one.

### CODE REVIEW

File: examples/run_lm_finetuning.py [hunk 1]
```diff
@@ -159,6 +152,20 @@ def _rotate_checkpoints(args, checkpoint_prefix, use_mtime=False):
...
+    if not args.save_total_limit:
+        return
+    if args.save_total_limit <= 0:
+        return
+
+    # Check if we should delete older checkpoint(s)
+    checkpoints_sorted = _sorted_checkpoints(args, checkpoint_prefix, use_mtime)
+    if len(checkpoints_sorted) <= args.save_total_limit:
+        return
+
```
Comment (julien-c) [inline examples/run_lm_finetuning.py]: Just refactored that function into two sub-units as I'm using the inner one elsewhere.

File: examples/run_lm_finetuning.py [hunk 2]
```diff
@@ -433,11 +439,16 @@ def main():
...
     parser.add_argument(
         "--output_dir",
-        default=None,
         type=str,
         required=True,
         help="The output directory where the model predictions and checkpoints will be written.",
     )
+    parser.add_argument(
+        "--model_type", type=str, required=True, help="The model architecture to be trained or fine-tuned.",
+    )
```
Comment (julien-c) [inline examples/run_lm_finetuning.py]: I passed that option as required as there's no reason to suppose want a BERT model by default anymore (but can revert)

File: examples/run_lm_finetuning.py [hunk 3]
```diff
@@ -627,26 +648,42 @@ def main():
...
+        model = model_class.from_pretrained(
+            args.model_name_or_path,
+            from_tf=bool(".ckpt" in args.model_name_or_path),
+            config=config,
+            cache_dir=args.cache_dir,
+        )
+    else:
+        logger.info("Training new model from scratch")
+        model = model_class(config=config)
+
```
Comment (julien-c) [inline examples/run_lm_finetuning.py]: This is the meat of the commit @LysandreJik @thomwolf. We are conditionally spawing new Config, Tokenizer, Model from pretrained, or from scratch.

File: examples/run_lm_finetuning.py [hunk 4]
```diff
@@ -493,9 +509,6 @@ def main():
     parser.add_argument(
         "--evaluate_during_training", action="store_true", help="Run evaluation during training at each logging step."
     )
-    parser.add_argument(
-        "--do_lower_case", action="store_true", help="Set this flag if you are using an uncased model."
-    )
```
Comment (julien-c) [inline examples/run_lm_finetuning.py]: At this point this should be an option of the tokenizer config, so I don't think we should leave that here. (Otherwise, all tokenizer options should also be there)

File: examples/run_lm_finetuning.py [hunk 5]
```diff
@@ -82,11 +83,11 @@
 
 
 class TextDataset(Dataset):
-    def __init__(self, tokenizer, args, file_path="train", block_size=512):
+    def __init__(self, tokenizer: PreTrainedTokenizer, args, file_path="train", block_size=512):
         assert os.path.isfile(file_path)
         directory, filename = os.path.split(file_path)
         cached_features_file = os.path.join(
-            directory, args.model_name_or_path + "_cached_lm_" + str(block_size) + "_" + filename
+            directory, args.model_type + "_cached_lm_" + str(block_size) + "_" + filename
```
Comment (LysandreJik) [inline examples/run_lm_finetuning.py]: This will break backward compatibility with users that have already tokenized their text into a file (not too bad), and won't allow multiple cache files for different checkpoints of the same model (annoying).

Couldn't we use the `model_name_or_path` and fallback to the `model_type` if training from scratch?

File: examples/run_lm_finetuning.py [hunk 6]
```diff
@@ -627,26 +648,42 @@ def main():
...
+        config = config_class()
+
+    if args.tokenizer_name:
+        tokenizer = tokenizer_class.from_pretrained(args.tokenizer_name, cache_dir=args.cache_dir)
+    elif args.model_name_or_path:
+        tokenizer = tokenizer_class.from_pretrained(args.model_name_or_path, cache_dir=args.cache_dir)
+    else:
+        logger.warning(
+            "You are instantiating a new {} tokenizer from scratch. Are you sure this is what you meant to do?"
+            "To specifiy a pretrained tokenizer name, use --tokenizer_name".format(tokenizer_class.__name__)
```
Comment (LysandreJik) [inline examples/run_lm_finetuning.py]: specifiy -> specify

Comment (LysandreJik) [inline examples/run_lm_finetuning.py]: As seen with you, this may be a bit misleading as it can be understood at "trains a tokenizer from scratch from the specified corpus", which is not the case.

Raising an error instead would probably be simpler and less error-prone for the users.

File: examples/run_lm_finetuning.py (hunk 5 above)
Comment (julien-c) [reply examples/run_lm_finetuning.py]: I guess we could, but honestly I don't even know if the previous behavior was very well defined, for instance if `model_name_or_path` is a path (like folder/folder/path)

For backward compatibility, I'm not sure a lot of users of an example script would upgrade the example script anyways (in the middle of using it/training models), what do you think?

File: examples/run_lm_finetuning.py (hunk 6 above)
Comment (julien-c) [reply examples/run_lm_finetuning.py]: Yeah, I agree. I'll remove this option and raise an error instead.

File: examples/run_lm_finetuning.py [hunk 7]
```diff
@@ -182,6 +202,8 @@ def mask_tokens(inputs: torch.Tensor, tokenizer: PreTrainedTokenizer, args) -> T
         tokenizer.get_special_tokens_mask(val, already_has_special_tokens=True) for val in labels.tolist()
     ]
     probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool), value=0.0)
+    padding_mask = labels.eq(tokenizer.pad_token_id)
+    probability_matrix.masked_fill_(padding_mask, value=0.0)
```
Comment (julien-c) [inline examples/run_lm_finetuning.py]: masking the padding tokens, cc @LysandreJik

File: examples/run_lm_finetuning.py [hunk 8]
```diff
@@ -120,13 +121,32 @@ def __getitem__(self, item):
         return torch.tensor(self.examples[item])
```
Comment (julien-c) [inline examples/run_lm_finetuning.py]: ^^ I did not add the last truncated example in the `TextDataset` (even though it should be handled by the padding and collation function below) @LysandreJik

**Comment by LysandreJik** [comment]
Looks good to me, feel free to merge it when you feel ready @julien-c

**Comment by shashankMadan-designEsthetics** [comment]
Yo! merge this shit up!!

**Comment by shashankMadan-designEsthetics** [comment]
Thanks, A Lot Guys your the best!
